src/Agent/NewRelic/Agent/Extensions/Providers/Wrapper/AzureFunction/InvokeFunctionAsyncWrapper.cs

```diff
@@ -114,64 +115,82 @@ internal class FunctionDetails
+            try
             {
-                // TODO: Needs null checks, optimization and possible caching of property accessors
-                // functionContext.FunctionDefinition.Parameters is an ImmutableArray<FunctionParameter>
-                var funcAsObj = (object)functionContext;
-                _functionDefinitionGetter ??= VisibilityBypasser.Instance.GeneratePropertyAccessor<object>(funcAsObj.GetType(), "FunctionDefinition");
-                var functionDefinition = _functionDefinitionGetter(funcAsObj);
+                FunctionName = functionContext.FunctionDefinition.Name;
+                InvocationId = functionContext.InvocationId;
 
-                _parametersGetter ??= VisibilityBypasser.Instance.GeneratePropertyAccessor<object>(functionDefinition.GetType(), "Parameters");
-                var parameters = _parametersGetter(functionDefinition) as IEnumerable;
-
-                var foundTrigger = false;
-                foreach (var parameter in parameters)
+                // cache the trigger by function name
+                if (!_functionTriggerCache.TryGetValue(FunctionName, out string trigger))
                 {
-                    // Properties is an IReadOnlyDictionary<string, object>
-                    _propertiesGetter ??= VisibilityBypasser.Instance.GeneratePropertyAccessor<IReadOnlyDictionary<string, object>>(parameter.GetType(), "Properties");
-                    var properties = _propertiesGetter(parameter);
+                    // functionContext.FunctionDefinition.Parameters is an ImmutableArray<FunctionParameter>
+                    var funcAsObj = (object)functionContext;
+                    _functionDefinitionGetter ??= VisibilityBypasser.Instance.GeneratePropertyAccessor<object>(funcAsObj.GetType(), "FunctionDefinition");
+                    var functionDefinition = _functionDefinitionGetter(funcAsObj);
 
-                    //foreach (var pair in properties)
-                    //{
-                    //    agent.Logger.Info($"{pair.Key}:{pair.Value.GetType().Name}");
-                    //}
+                    _parametersGetter ??= VisibilityBypasser.Instance.GeneratePropertyAccessor<object>(functionDefinition.GetType(), "Parameters");
+                    var parameters = _parametersGetter(functionDefinition) as IEnumerable;
 
-                    foreach (var propVal in properties.Values)
+                    // Trigger is normally the first parameter, but we'll check all parameters to be sure.
+                    var foundTrigger = false;
+                    foreach (var parameter in parameters)
                     {
-                        if (propVal.GetType().Name.Contains("Trigger"))
+                        // Properties is an IReadOnlyDictionary<string, object>
+                        _propertiesGetter ??= VisibilityBypasser.Instance.GeneratePropertyAccessor<IReadOnlyDictionary<string, object>>(parameter.GetType(), "Properties");
```

Comment:
If the trigger isn't the first parameter, won't `_propertiesGetter` always point to the first parameter?

Reply:
All the parameters should be of the FunctionParameter type, so I would expect the getter to work for each parameter.

Reply:
Ah ok, that makes sense